sound/soc/sof/probe.c

```diff
@@ -0,0 +1,205 @@
+
+	ret = sof_ipc_tx_message(sdev->ipc, msg->hdr.cmd, msg, msg->hdr.size,
+			&reply, sizeof(reply));
+	kfree(msg);
+	return ret;
+}
+EXPORT_SYMBOL(sof_ipc_probe_init);
+
+int sof_ipc_probe_deinit(struct snd_sof_dev *sdev)
+{
+	struct sof_ipc_cmd_hdr msg;
+	struct sof_ipc_reply reply;
+
+	msg.size = sizeof(msg);
+	msg.cmd = SOF_IPC_GLB_PROBE | SOF_IPC_PROBE_DEINIT;
+
+	return sof_ipc_tx_message(sdev->ipc, msg.cmd, &msg, msg.size,
+			&reply, sizeof(reply));
+}
+EXPORT_SYMBOL(sof_ipc_probe_deinit);
+
+static int sof_ipc_probe_get(struct snd_sof_dev *sdev, unsigned int cmd,
+		void **params, size_t *bytes)
+{
+	struct sof_ipc_probe_get_params msg = {0};
+	struct sof_ipc_probe_get_params *reply;
+	size_t size;
+	int ret;
+
+	*params = NULL;
+	*bytes = 0;
+
+	reply = kzalloc(SOF_IPC_MSG_MAX_SIZE, GFP_KERNEL);
+	if (!reply)
+		return -ENOMEM;
+	msg.rhdr.hdr.size = sizeof(msg);
+	msg.rhdr.hdr.cmd = SOF_IPC_GLB_PROBE | cmd;
+
+	ret = sof_ipc_tx_message(sdev->ipc, msg.rhdr.hdr.cmd, &msg,
+			msg.rhdr.hdr.size, reply, SOF_IPC_MSG_MAX_SIZE);
+	if (ret < 0 || reply->rhdr.error < 0)
+		goto exit;
+
+	size = reply->rhdr.hdr.size -= sizeof(*reply);
+	if (size <= 0)
+		goto exit;
+
+	*params = kzalloc(size, GFP_KERNEL);
+	if (!*params) {
+		ret = -ENOMEM;
+		goto exit;
+	}
+	memcpy(*params, &reply->dma[0], size);
+	*bytes = size;
+
+exit:
+	kfree(reply);
+	return ret;
+}
+
+int sof_ipc_probe_get_dma(struct snd_sof_dev *sdev,
+		struct sof_probe_dma **dma, size_t *num_dma)
+{
+	int ret;
+
+	ret = sof_ipc_probe_get(sdev, SOF_IPC_PROBE_DMA_GET,
+			(void **)dma, num_dma);
+	*num_dma /= sizeof(**dma);
+	return ret;
+}
+EXPORT_SYMBOL(sof_ipc_probe_get_dma);
+
+int sof_ipc_probe_dma_attach(struct snd_sof_dev *sdev,
+		struct sof_probe_dma *dma, size_t num_dma)
+{
+	struct sof_ipc_probe_dma_set_params *msg;
+	struct sof_ipc_reply reply;
+	size_t bytes = sizeof(*dma) * num_dma;
+	int ret;
+
+	msg = kzalloc(sizeof(*msg) + bytes, GFP_KERNEL);
+	if (!msg)
+		return -ENOMEM;
+	msg->hdr.size = sizeof(*msg) + bytes;
+	msg->hdr.cmd = SOF_IPC_GLB_PROBE | SOF_IPC_PROBE_DMA_SET;
+	memcpy(&msg->dma[0], dma, bytes);
+
+	ret = sof_ipc_tx_message(sdev->ipc, msg->hdr.cmd, msg, msg->hdr.size,
+			&reply, sizeof(reply));
+	kfree(msg);
+	return ret;
+}
+EXPORT_SYMBOL(sof_ipc_probe_dma_attach);
+
+int sof_ipc_probe_dma_detach(struct snd_sof_dev *sdev,
+		unsigned int *stream_tag, size_t num_stream_tag)
+{
+	struct sof_ipc_probe_dma_detach_params *msg;
+	struct sof_ipc_reply reply;
+	size_t bytes = sizeof(*stream_tag) * num_stream_tag;
+	int ret;
+
+	msg = kzalloc(sizeof(*msg) + bytes, GFP_KERNEL);
+	if (!msg)
+		return -ENOMEM;
+	msg->hdr.size = sizeof(*msg) + bytes;
+	msg->hdr.cmd = SOF_IPC_GLB_PROBE | SOF_IPC_PROBE_DMA_DETACH;
+	memcpy(&msg->stream_tag[0], stream_tag, bytes);
+
+	ret = sof_ipc_tx_message(sdev->ipc, msg->hdr.cmd, msg, msg->hdr.size,
+			&reply, sizeof(reply));
+	kfree(msg);
+	return ret;
+}
+EXPORT_SYMBOL(sof_ipc_probe_dma_detach);
```

Comment:
I think we should combine the attach/detach routines into one and pass the command as an argument

Reply:
hmm I Just saw that the msg type is different for these. so I guess its easier to just use different methods.

Reply:
Hmm, I did combine both getters as IPC structure is shared, yet is this case we are dealing with different argument list. msg pointers are also of different type. Are you sure about this?